File: server/internal/store/list/archive/find.go
package archive

import (
	"context"
	"time"

	"database/sql"

	"github.com/emonadeo/gdol/internal/openapi"
	"github.com/emonadeo/gdol/internal/util"
	"github.com/lib/pq"
)

const qryFind = `
SELECT
    list_archive.timestamp,
    levels.id,
    levels.name,
    levels.gd_id,
    levels.video,
    levels.requirement,
    users.id AS user_id,
    users.name AS user_name,
    users.nationality AS user_nationality,
    users.discord_id AS user_discord_id,
    verifier.id AS verifier_id,
    verifier.name AS verifier_name,
    verifier.nationality AS verifier_nationality,
    verifier.discord_id AS verifier_discord_id,
    array_remove(array_agg(creators.id), NULL) :: bigint [ ] AS creator_ids,
    array_remove(array_agg(creators.name), NULL) :: text [ ] AS creator_names,
    array_agg(creators.nationality) AS creator_nationalities,
    array_agg(creators.discord_id) AS creator_discord_ids
FROM
    list_archive
    JOIN levels ON levels.id = list_archive.level_id
    JOIN users ON users.id = levels.user_id
    JOIN users verifier ON verifier.id = levels.verifier_id
    LEFT JOIN user_created_level ON user_created_level.level_id = list_archive.level_id
    LEFT JOIN users creators ON creators.id = user_created_level.user_id
GROUP BY
    list_archive.timestamp,
    levels.id,
    users.id,
    verifier.id;
`

func (a Archive) Find(ctx context.Context) ([]openapi.LevelArchived, error) {
	// Cannot use sqlc because of https://github.com/kyleconroy/sqlc/issues/185
	rows, err := a.DB.QueryContext(ctx, qryFind)
	if err != nil {
		return nil, err
	}

	levels := []openapi.LevelArchived{}

	for rows.Next() {
		var timestamp time.Time
		var id int64
		var name string
		var gdId sql.NullInt64
		var video sql.NullString
		var requirement sql.NullInt16
		var userId int64
		var userName string
		var userNationality sql.NullString
		var userDiscordId sql.NullString
		var verifierId int64
		var verifierName string
		var verifierNationality sql.NullString
		var verifierDiscordId sql.NullString
		var creatorIds []int64
		var creatorNames []string
		var creatorNationalities []sql.NullString
		var creatorDiscordIds []sql.NullString

		rows.Scan(
			&timestamp,
			&id,
			&name,
			&gdId,
			&video,
			&requirement,
			&userId,
			&userName,
			&userNationality,
			&userDiscordId,
			&verifierId,
			&verifierName,
			&verifierNationality,
			&verifierDiscordId,
			pq.Array(&creatorIds),
			pq.Array(&creatorNames),
			pq.Array(&creatorNationalities),
			pq.Array(&creatorDiscordIds),
		)

		creators := []openapi.User{}
		for i, id := range creatorIds {
			creators = append(creators, openapi.User{
				Id:          id,
				Name:        creatorNames[i],
				Nationality: util.NullString(creatorNationalities[i]),
			})
		}

		levels = append(levels, openapi.LevelArchived{
			Timestamp:   timestamp,
			Id:          id,
			Name:        name,
			GdId:        util.NullInt64(gdId),
			Video:       util.NullString(video),
			Requirement: util.NullInt16(requirement),
			User: openapi.User{
				Id:          userId,
				Name:        userName,
				Nationality: util.NullString(userNationality),
			},
			Verifier: openapi.User{
				Id:          verifierId,
				Name:        verifierName,
				Nationality: util.NullString(verifierNationality),
			},
			Creators: creators,
		})
	}
	return levels, nil
}

File: store/list/find.ts
import { Database } from '../../deps.ts';
import { Level } from '../../types/mod.ts';

const query = await (await fetch(new URL('./find.sql', import.meta.url))).text();

type RawLevel = {
	level_id: number;
	level_name: string;
	level_gd_id: number | null;
	level_video: string | null;
	level_requirement: number | null;
	user_id: number;
	user_name: string;
	user_nationality: string | null;
	verifier_id: number;
	verifier_name: string;
	verifier_nationality: string | null;
	creator_ids: string;
	creator_names: string;
	creator_nationalities: string;
};

function levelFromRaw(level: RawLevel): Level {
	const creator_ids = JSON.parse(level.creator_ids) as number[];
	const creator_names = JSON.parse(level.creator_names) as string[];
	const creator_nationalities = JSON.parse(level.creator_nationalities) as Array<string | null>;

	return {
		id: level.level_id,
		name: level.level_name,
		gd_id: level.level_gd_id ?? undefined,
		video: level.level_video ?? undefined,
		requirement: level.level_requirement ?? undefined,
		user: {
			id: level.user_id,
			name: level.user_name,
			nationality: level.user_nationality ?? undefined,
		},
		verifier: {
			id: level.verifier_id,
			name: level.verifier_name,
			nationality: level.verifier_nationality ?? undefined,
		},
		creators: creator_ids.map((creator_id, i) => ({
			id: creator_id,
			name: creator_names.at(i) as string,
			nationality: creator_nationalities.at(i) ?? undefined,
		})),
	};
}

export function find(db: Database): Level[] {
	return db.prepare(query).all<RawLevel>().map(levelFromRaw);
}
